x-pack/plugins/fleet/server/plugin.ts

```diff
@@ -332,8 +333,35 @@ export class FleetPlugin
 
     this.telemetryEventsSender.start(plugins.telemetry, core);
 
+    const logger = appContextService.getLogger();
+
+    const fleetSetupPromise = async () => {
+      try {
+        logger.info('Beginning fleet setup');
+
+        const { nonFatalErrors } = await setupFleet(
+          new SavedObjectsClient(core.savedObjects.createInternalRepository()),
+          core.elasticsearch.client.asInternalUser
+        );
+
+        if (nonFatalErrors.length > 0) {
+          logger.info('Encountered non fatal errors during Fleet setup');
+          formatNonFatalErrors(nonFatalErrors).forEach((error) =>
+            logger.info(JSON.stringify(error))
+          );
+        }
+
+        logger.info('Fleet setup completed');
+      } catch (error) {
+        logger.warn('Fleet setup failed');
+        logger.warn(error);
+      }
+    };
+
+    fleetSetupPromise();
+
     return {
-      fleetSetupCompleted: () => Promise.resolve(),
+      fleetSetupCompleted: fleetSetupPromise,
```

Comment:
I think with that approach every plugin consuming `fleetSetupCompleted` is going to trigger a new call to setup no?

Reply:
Yeah you are right - we need to track whether setup is already completed or not in this class.

Reply:
is already completed or already started. (probably one of the last thing to fix on that PR, I played a little with it locally and it seems to works well :) )

Reply:
Just added some status tracking to fix possible duplicate calls of setup in 95475d4c020701b9ce6346287299271849eada3b

Reply:
I think if it's pending we should probably return the original promise, so the code calling it will still wait for the setup to complete